gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -636,7 +646,9 @@ private int getNumDatasetsVerifiedTogether() {
 
   private void setAllDatasetStatesToVerified() {
     for (Dataset dataset : this.datasets) {
-      dataset.setState(VERIFIED);
+      if (dataset.state() == UNVERIFIED) {
```

Comment:
This is another good indicator that you actually need to use a `AtomicReference` if you need to guard `state` since this check-then-act can be done in a single function call.